src/main/java/org/jenkinsci/plugins/workflow/steps/durable_task/DurableTaskStep.java

```diff
@@ -353,17 +484,111 @@ private void check() {
+                getContext().onSuccess(returnStatus ? exitCode : returnStdout ? new String(output, StandardCharsets.UTF_8) : null);
+            } else {
+                if (returnStdout) {
+                    listener().getLogger().write(output); // diagnostic
+                }
+                getContext().onFailure(new AbortException("script returned exit code " + exitCode));
+            }
+        }
+
+        // ditto
+        @Override public void problem(Exception x) {
+            Functions.printStackTrace(x, listener().getLogger());
+            // note that if there is _also_ a problem in the master-side logger, PrintStream will mask it
+        }
+
         @Override public void onResume() {
-            setupTimer();
+            ws = null; // find it from scratch please
+            setupTimer(MIN_RECURRENCE_PERIOD);
+            // In watch mode, we will quickly enter the check.
+            // Then in getWorkspace when ws == null we will start a watch and go back to sleep.
         }
 
-        private void setupTimer() {
-            recurrencePeriod = MIN_RECURRENCE_PERIOD;
+        private void setupTimer(long initialRecurrencePeriod) {
+            recurrencePeriod = initialRecurrencePeriod;
             task = THREAD_POOL.schedule(this, recurrencePeriod, TimeUnit.MILLISECONDS);
         }
 
         private static final long serialVersionUID = 1L;
 
     }
 
+    private static class HandlerImpl extends Handler {
+
+        private static final Field printStreamDelegate;
+        static {
+            try {
+                printStreamDelegate = FilterOutputStream.class.getDeclaredField("out");
+            } catch (NoSuchFieldException x) {
+                // Defined in Java Platform and protected, so should not happen.
+                throw new ExceptionInInitializerError(x);
+            }
+            printStreamDelegate.setAccessible(true);
+        }
+
+        private static final long serialVersionUID = 1L;
+
+        private final ExecutionRemotable execution;
+        private final TaskListener listener;
+
+        HandlerImpl(Execution execution, FilePath workspace, TaskListener listener) {
+            this.execution = workspace.getChannel().export(ExecutionRemotable.class, execution);
+            this.listener = listener;
+        }
+
+        @Override public void output(InputStream stream) throws Exception {
+            PrintStream ps = listener.getLogger();
+            OutputStream os;
+            if (ps.getClass() == PrintStream.class) {
+                // Try to extract the underlying stream, since swallowing exceptions is undesirable and PrintStream.checkError is useless.
+                os = (OutputStream) printStreamDelegate.get(ps);
+            } else {
+                // A subclass. Who knows why, but trust any write(…) overrides it may have.
+                os = ps;
+            }
+            try {
+                IOUtils.copy(stream, os);
```

Comment:
It looks like all `PrintStream#write` overloads synchronize on `this` and null-check `out` before writing (since it is nulled out once the stream is closed). Do we need to do that here as well to avoid breaking `PrintStream` invariants? The usage here should always be from the same thread IIUC, but do other threads have access to `listener.getLogger()`?

Reply:
Agreed, that sounds safer.